crates/exex/exex/src/wal/storage.rs

```diff
@@ -0,0 +1,217 @@
+use std::{
+    fs::File,
+    io::{BufRead, BufReader, BufWriter, Read, Seek, SeekFrom, Write},
+    ops::ControlFlow,
+    path::{Path, PathBuf},
+};
+
+use reth_exex_types::ExExNotification;
+
+/// The underlying WAL storage backed by a file.
+///
+/// Each notification is written without any delimiters and structured as follows:
+/// ```text
+/// +--------------------------+----------------------------------+
+/// | little endian u32 length | MessagePack-encoded notification |
+/// +--------------------------+----------------------------------+
+/// ```
+/// The length is the length of the MessagePack-encoded notification in bytes.
+#[derive(Debug)]
+pub(super) struct Storage {
+    /// The path to the WAL file.
+    path: PathBuf,
+    /// The file handle of the WAL file.
+    file: File,
+}
+
+impl Storage {
+    /// Creates a new instance of [`Storage`] backed by the file at the given path and creates
+    /// it doesn't exist.
+    pub(super) fn new(path: impl AsRef<Path>) -> eyre::Result<Self> {
+        Ok(Self { path: path.as_ref().to_path_buf(), file: Self::open_file(&path)? })
+    }
+
+    /// Opens the file at the given path and creates it if it doesn't exist.
+    pub(super) fn open_file(path: impl AsRef<Path>) -> std::io::Result<File> {
+        File::options().read(true).write(true).create(true).truncate(false).open(path)
+    }
+
+    /// Returns the length of the underlying file in bytes.
+    pub(super) fn bytes_len(&self) -> std::io::Result<u64> {
+        Ok(self.file.metadata()?.len())
+    }
+
+    /// Truncates the underlying file from the given byte offset (inclusive) to the end of the file.
+    ///
+    /// 1. Creates a new file and copies all notifications starting from the offset (inclusive) to
+    ///    the end of the original file.
+    /// 2. Renames the new file to the original file.
+    ///
+    /// # Returns
+    ///
+    /// The old and new underlying file sizes in bytes
+    pub(super) fn truncate_from_offset(&mut self, offset: u64) -> eyre::Result<(u64, u64)> {
```

Comment:
so finalizing a block means we must copy and reinitialize the entire file?
with > 64 blocks (more with reorged blocks) this isn't very cheap I suppose

what happens to the offsets that the cache tracks?